File: schemas/video-conferencing/drizzle/recordings.ts
// recordings: Captured audio/video recordings of meetings.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, integer, bigint, timestamp, index } from "drizzle-orm/pg-core";
import { meetings } from "./meetings";
import { users } from "./users";
import { files } from "./files";

export const recordingTypeEnum = pgEnum("recording_type", ["composite", "audio_only", "video_only", "screen_share"]);
export const recordingStatusEnum = pgEnum("recording_status", ["recording", "processing", "ready", "failed", "deleted"]);

export const recordings = pgTable(
  "recordings",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    meetingId: uuid("meeting_id").notNull().references(() => meetings.id, { onDelete: "cascade" }),
    fileId: uuid("file_id").references(() => files.id, { onDelete: "set null" }),
    type: recordingTypeEnum("type").notNull().default("composite"),
    status: recordingStatusEnum("status").notNull().default("recording"),
    durationSeconds: integer("duration_seconds"),
    fileSize: bigint("file_size", { mode: "number" }),
    startedAt: timestamp("started_at", { withTimezone: true }).notNull().defaultNow(),
    endedAt: timestamp("ended_at", { withTimezone: true }),
    startedBy: uuid("started_by").references(() => users.id, { onDelete: "set null" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_recordings_meeting_id").on(table.meetingId),
    index("idx_recordings_status").on(table.status),
    index("idx_recordings_started_by").on(table.startedBy),
  ]
);

File: schemas/video-conferencing/spacetimedb/recordings.rs
// recordings: captured audio/video recordings of meeting sessions.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum RecordingType {
    Composite, // type: String
    AudioOnly,
    VideoOnly,
    ScreenShare,
}

#[derive(SpacetimeType, Clone)]
pub enum RecordingStatus {
    Recording, // type: String
    Processing,
    Ready,
    Failed,
    Deleted,
}

#[spacetimedb::table(name = recordings, public)]
pub struct Recording {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub meeting_id: String, // UUID — FK → meetings.id (cascade delete)
    pub file_id: Option<String>, // UUID — FK → files.id (set null on delete)
    pub recording_type: RecordingType,
    #[index(btree)]
    pub recording_status: RecordingStatus,
    pub duration_seconds: Option<i32>,
    pub file_size: Option<i64>,
    pub started_at: Timestamp,
    pub ended_at: Option<Timestamp>,
    #[index(btree)]
    pub started_by: Option<String>, // UUID — FK → users.id (set null on delete)
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
